### PR TITLE
rewrite solution

### PR DESCRIPTION
**Pull request by KrisMur**
[DEMO LINK](https://KrisMur.github.io/react_todo-app/)

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,93 +1,219 @@
...
+
+  useEffect(() => {
+    localStorage.setItem('todos', JSON.stringify(todos));
+  }, [todos]);
+
+  const [idToDelete, setIdToDelete] = useState(0);
+  const [stateOfTodo, setStateOfTodo]
+    = useState<Todo>({ id: 0, state: TodoState.COMPLETED });
+  const [titleOfTodo, setTitleOfTodo]
+    = useState<Todo>({ id: 0, title: '', state: TodoState.ACTIVE });
```
Comment (jstmpelowycz) [inline src/App.tsx:30]: you'd rather put single properties to state, not complete todo objects

File: src/App.tsx [hunk 2]
```diff
@@ -1,93 +1,219 @@
...
+      return prev;
+    };
+
+    const changedTodos = todos.reduce(changeTodo, []);
+
+    setTodos(changedTodos);
+  };
+
+  const countOfActiveTodos = todos.filter(todo => (
+    todo.state !== TodoState.COMPLETED)).length;
```
Comment (jstmpelowycz) [inline src/App.tsx:71]:
```suggestion
  const countOfActiveTodos = todos.filter(todo => (
    todo.state !== TodoState.COMPLETED)
  ).length;
```

File: src/App.tsx (hunk 1 above)
Comment (jstmpelowycz) [reply src/App.tsx:30]: morever, names of states do not match what you keep inside

File: src/App.tsx [hunk 3]
```diff
@@ -1,93 +1,219 @@
...
+          <Route
+            path="/completed"
+            element={(
+              <TodoList
+                todos={todos.filter(todo => todo.state === TodoState.COMPLETED)}
+                setIdToDelete={setIdToDelete}
+                setStateOfTodo={setStateOfTodo}
+                setTitleOfTodo={setTitleOfTodo}
+              />
+            )}
```
Comment (jstmpelowycz) [inline src/App.tsx:156]: DRY

File: src/App.tsx [hunk 4]
```diff
@@ -1,93 +1,219 @@
-/* eslint-disable jsx-a11y/control-has-associated-label */
```
Comment (jstmpelowycz) [inline src/App.tsx:1]: too much logic for a file, split into several ones

File: src/components/Header.tsx [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+type Props = {
+  todos: Todo[],
+  setTodos: (todos: Todo[]) => void,
+};
+
+export const Header: React.FC<Props> = ({ todos, setTodos }) => {
+  const [query, setQuery] = useState('');
+
+  const handler = (event: React.FormEvent<HTMLFormElement>) => {
```
Comment (jstmpelowycz) [inline src/components/Header.tsx:13]: handler of what?)

File: src/components/ToDoList.tsx [hunk 6]
```diff
@@ -0,0 +1,29 @@
+import { Todo } from '../types/ToDo';
+import { TodoElement } from './ToDoElement';
+
+type Props = {
+  todos: Todo[],
+  setIdToDelete: (todoId: number) => void,
+  setStateOfTodo: (arg0: Todo) => void,
+  setTitleOfTodo: (todoTitle: Todo) => void,
```
Comment (jstmpelowycz) [inline src/components/ToDoList.tsx:8]:
```
const handleSomething = (/*params*/) => ...;
...

<SomeComponent onSomething={handleSomething} />
...

interface Props {
  onSomethings: (/*params*/) => void;
}
```

Comment (jstmpelowycz) [reply src/components/ToDoList.tsx:8]: stick to proper naming, it's indeed important

**Review by jstmpelowycz — CHANGES_REQUESTED** [review]
Screenshot 2022-08-16 at 10 02 05
good job! make sure you resolve the issues, and also you have to fix that `0`

File: src/UseLocalStorage.ts [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+    }
+  });
+
+  const save = (newValue: T) => {
+    setValue(newValue);
+    localStorage.setItem(key, JSON.stringify(value));
+  };
+
+  return [value, save] as const;
+}
```
Comment (Olexandr6) [inline src/UseLocalStorage.ts:18]: 👍

**Review by Olexandr6 — APPROVED** [review]
Looks good🔥

**Pull request closed without merging**